model/task_queue_service_dependency.go

```diff
@@ -253,14 +253,20 @@ func (d *basicCachedDAGDispatcherImpl) FindNextTask(ctx context.Context, spec Ta
 
 	settings := evergreen.GetEnvironment().Settings()
 	dependencyCaches := make(map[string]task.Task)
-	for i := range d.sorted {
-		node := d.sorted[i]
+	d.mu.RLock()
+	sorted := make([]graph.Node, len(d.sorted))
+	copy(sorted, d.sorted)
```

Comment:
Using a copy of the sorted list should be fine because the item retrieval on L268 already handles the item being nil, and `node.ID()` just yields a task ID, which is unique.

Reply:
The potential staleness of the sorted slice should only potentially affect efficiency, not correctness, because it is mitigated by real-time fetch and write operations:
-  if an item in the stale copy has been removed from the real slice, [getItemByNodeID](https://github.com/evergreen-ci/evergreen/pull/8311/files#diff-b843eb16a854431304da1fbc4589faa91d2464dcc0911a02993da858307b6242R268) will return nil and the loop will skip it, and it also ensures we're working with the current state of the item
- the lock in [tryMarkItemDispatched](https://github.com/evergreen-ci/evergreen/pull/8311/files#diff-b843eb16a854431304da1fbc4589faa91d2464dcc0911a02993da858307b6242R453) ensures that even if two goroutines are working on copies of the same slice and reach the same item, only one will successfully mark it dispatched and return it because the other goroutine will recognize that the item has been marked dispatched
- if the stale copy is missing new items those items won't be processed in this loop, but they'll be available in future calls.

This granular locking is definitely more complicated than what it was before, but given the amount of DB calls the dispatching logic has, it should yield solid performance benefits. I'll write a big comment. FWIW average next task latency has dropped from [450ms](https://ui.honeycomb.io/mongodb-4b/environments/production/datasets/evergreen/result/3kXgR2aWo6V?hideCompare) to [150ms](https://ui.honeycomb.io/mongodb-4b/environments/production/datasets/evergreen/result/8xgbWxkDvCP?hideCompare) since we've deployed this granular locking, which is promising.